pkg/agent/controller/networkpolicy/reconciler.go

```diff
@@ -541,18 +542,31 @@ func (r *reconciler) computeOFRulesForAdd(rule *CompletedRule, ofPriority *uint1
 		return ofRuleByServicesMap, lastRealized
 	}
 	if rule.Direction == v1beta2.DirectionIn {
+		isRuleAppliedToService := isRuleAppliedToService(rule.TargetMembers)
 		// Addresses got from source GroupMembers' IPs.
 		from1 := groupMembersToOFAddresses(rule.FromAddresses)
 		// Get addresses that in From IPBlock but not in Except IPBlocks.
-		from2 := ipBlocksToOFAddresses(rule.From.IPBlocks, r.ipv4Enabled, r.ipv6Enabled)
+		from2 := ipBlocksToOFAddresses(rule.From.IPBlocks, r.ipv4Enabled, r.ipv6Enabled, isRuleAppliedToService)
 		membersByServicesMap, servicesMap := groupMembersByServices(rule.Services, rule.TargetMembers)
 		for svcKey, members := range membersByServicesMap {
-			ofPorts := r.getOFPorts(members)
-			lastRealized.podOFPorts[svcKey] = ofPorts
+			toAddresses := make([]types.Address, 0)
+			if isRuleAppliedToService {
+				addressSet := sets.NewInt64()
+				for _, member := range members.Items() {
+					curAddresses, curAddressSet := r.svcRefToOFAddresses(*member.Service)
+					toAddresses = append(toAddresses, curAddresses...)
+					addressSet = addressSet.Union(curAddressSet)
```

Comment:
I think it's less efficient to let `svcRefToOFAddresses` return a set as calling `Union` copies two sets every time.
If it justs return a slice of int64, we could just insert the returned slice to the set in the loop:
```
curAddresses, curGroupIDs := r.svcRefToOFAddresses(*member.Service)
toAddresses = append(toAddresses, curAddresses...)
addressSet.Insert(curGroupIDs...)
```

Reply:
Good one. Updated with more refactor on this part.